File: solutions/challenge2/ch2.py
def swap_nibbles(n):
    """Swap nibbles of a byte.

    Args:
        n (int)

    Return:
        a byte with its halves swapped."""

    # Get low nibble (low 4 bits)
    lo = n & 0xF

    # Get high nibble (high 4 bits)
    hi = (n >> 4) & 0xF

    return (lo << 4) | hi


if __name__ == '__main__':
    for i in range(0, 16):
        for j in range(0, 16):
            assert(swap_nibbles(i * 16 + j) == (j << 4) | i)

File: solutions/challenge2/ch2.c
/* Compile with:

   gcc ch2.c -std=c99 -o ch2
*/

#include <stdlib.h>
#include <assert.h>

unsigned swap_nibbles(unsigned n)
{
    /* Get low nibble (low 4 bits) */
    unsigned lo = n & 0xF;

    /* Get high nibble (high 4 bits) */
    unsigned hi = (n >> 4) & 0xF;

    return (lo << 4) | hi;
}

int main()
{
    for (int i = 0; i < 16; i++)
        for (int j = 0; j < 16; j++)
            assert(swap_nibbles(i * 16 + j) == (j << 4) | i);
}
